docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+
+```protobuf
+syntax = "proto3";
+
+package job.v1;
+
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/kralicky/jobserver/pkg/apis/job/v1";
+
+service Job {
+  // Starts a new job, and returns its id if it was started successfully.
+  //
+  // The job will be run asynchronously; this method does not wait for it
+  // to complete.
+  //
+  // If resource limits are provided, the job will be run in its own cgroup
+  // with all the given limits applied. Any combination of the allowed limit
+  // types may be provided; it is not necessary to set all of them.
+  rpc Start(JobSpec) returns (JobId);
+
+  // Stops a running job. Once stopped, this method will wait until the job has
+  // completed before returning.
+  //
+  // This will first attempt to stop the job's process using SIGTERM, but if the
+  // process does not exit within a short grace period, it will be forcefully
+  // killed with SIGKILL.
+  //
+  // A job must be in the Running state to be stopped. If the job is in any
+  // other state, this returns a FailedPrecondition error.
+  rpc Stop(JobId) returns (google.protobuf.Empty);
+
+  // Returns the status of an existing job.
+  //
+  // If the job is completed, the exit code and detailed wait status will be
+  // present in the response. Jobs stopped by the user with the Stop() method
+  // may not have an exit code, depending if the process was able to exit
+  // cleanly within the grace period. Jobs that failed to start, or were
+  // otherwise killed by the system, will not have an exit code, but may have
+  // a detailed wait status indicating the reason for failure.
+  rpc Status(JobId) returns (JobStatus);
+
+  // Returns a list of all job IDs that are currently known to the server.
+  //
+  // All jobs, regardless of state, are included. No guarantees are made about
+  // the order of ids in the list.
+  rpc List(google.protobuf.Empty) returns (JobIdList);
+
+  // Streams the combined stdout and stderr output of a running or completed job.
+  // The output always starts from the beginning of process execution. Jobs that
+  // were stopped by the user with the Stop() method will have output up to the
+  // time they were stopped.
+  //
+  // If 'follow' is set to true, the output will continue to be be streamed in
+  // real-time until the job completes. Otherwise, only the currently accumulated
+  // output will be written, and the stream will be closed even if the job is
+  // still running.
+  //
+  // If the job is already completed, the full output of the job will be
+  // written to the stream, after which the stream will be closed.
+  rpc Output(OutputRequest) returns (stream ProcessOutput);
+}
+
+// JobSpec describes a command to be run, along with optional resource limits
+// that should be applied to the command's process.
+message JobSpec {
+  CommandSpec    command = 1;
+  ResourceLimits limits  = 2;
+}
+
+message JobId {
+  string id = 1;
```

Comment:
How is the job id going to be generated?

Reply:
I'll go with UUIDs probably.